src/main/java/de/unistuttgart/overworldbackend/service/PlayerTaskStatisticService.java

```diff
@@ -0,0 +1,202 @@
+      minigameTask.get().getId(),
+      lecture.getId(),
+      playerstatistic.getId()
+    );
+    final PlayerTaskStatistic currentPlayerTaskStatistic;
+    if (playerTaskStatistic.isEmpty()) {
+      final PlayerTaskStatistic newPlayerTaskStatistic = new PlayerTaskStatistic();
+      newPlayerTaskStatistic.setPlayerstatistic(playerstatistic);
+      newPlayerTaskStatistic.setMinigameTask(minigameTask.get());
+      newPlayerTaskStatistic.setLecture(lecture);
+      currentPlayerTaskStatistic = playerTaskStatisticRepository.save(newPlayerTaskStatistic);
+    } else {
+      currentPlayerTaskStatistic = playerTaskStatistic.get();
+    }
+
+    long gainedKnowledge = calculateKnowledge(data.getScore(), currentPlayerTaskStatistic.getHighscore());
+
+    if (data.getScore() > currentPlayerTaskStatistic.getHighscore()) {
+      currentPlayerTaskStatistic.setHighscore(data.getScore());
+    }
+    if (!currentPlayerTaskStatistic.isCompleted()) {
+      currentPlayerTaskStatistic.setCompleted(checkCompleted(data.getScore()));
+    }
+
+    logData(data, lecture, currentPlayerTaskStatistic, gainedKnowledge);
+
+    calculateCompletedDungeons(playerstatistic);
+
+    //TODO:calculate completed dungeons and unlocked areas
+
+    playerstatistic.addKnowledge(gainedKnowledge);
+    playerstatisticRepository.save(playerstatistic);
+
+    return playerTaskStatisticMapper.playerTaskStatisticToPlayerTaskStatisticDTO(
+      playerTaskStatisticRepository.save(currentPlayerTaskStatistic)
+    );
+  }
+
+  private void calculateCompletedDungeons(final Playerstatistic playerstatistic) {
+    final List<PlayerTaskStatistic> playerTaskStatistics = playerTaskStatisticRepository.findPlayerTaskStatisticByPlayerstatisticId(
+      playerstatistic.getId()
+    );
+    final List<World> worlds = playerstatistic.getLecture().getWorlds();
+    final List<Dungeon> completedDungeons = new ArrayList<>();
+    for (final World world : worlds) {
+      for (final Dungeon dungeon : world.getDungeons()) {
+        boolean completed = true;
+        for (final MinigameTask minigameTask : dungeon.getMinigameTasks()) {
+          final Optional<PlayerTaskStatistic> currentStatistic = playerTaskStatistics
+            .stream()
+            .filter(filteredPlayerstatistic -> filteredPlayerstatistic.getMinigameTask().equals(minigameTask))
+            .findAny();
+          if (currentStatistic.isEmpty() || !currentStatistic.get().isCompleted()) {
+            completed = false;
+          }
+        }
+        if (completed) {
+          completedDungeons.add(dungeon);
+        }
+      }
+    }
+    final List<AreaLocation> completedDungeonLocations = new ArrayList<>();
+    for (final Dungeon dungeon : completedDungeons) {
+      completedDungeonLocations.add(new AreaLocation(dungeon.getWorld(), dungeon));
+    }
+    playerstatistic.setCompletedDungeons(completedDungeonLocations);
```

Comment:
```suggestion
    final List<PlayerTaskStatistic> playerTaskStatistics = playerTaskStatisticRepository.findPlayerTaskStatisticByPlayerstatisticId(
      playerstatistic.getId()
    );
    playerstatistic.setCompletedDungeons(playerstatistic.getLecture().getWorlds().parallelStream().map(World::getDungeons).flatMap(Collection::parallelStream).filter(dungeon -> dungeon.getMinigameTasks().parallelStream().allMatch(minigameTask -> playerTaskStatistics.parallelStream().filter(taskStatistic -> taskStatistic.getMinigameTask().equals(minigameTask)).allMatch(PlayerTaskStatistic::isCompleted))).map(dungeon -> new AreaLocation(dungeon.getWorld(), dungeon)).toList());
```
Holy fucking shit, I think I've finally been able to reduce this monstrosity into a few far more understandable streams...
Let's hope I didn't mess up any parentheses...